packages/runtime-core/__tests__/apiWatch.spec.ts

```diff
@@ -1290,6 +1290,33 @@ describe('api: watch', () => {
     expect(spy).toHaveBeenCalledTimes(2)
   })
 
+  test('handle nested watcher stop properly', () => {
+    let instance: ComponentInternalInstance
+    const Comp = {
+      setup() {
+        instance = getCurrentInstance()!
+        watch(
+          () => 1,
+          (val, oldVal, onCleanup) => {
+            const stop = watch(
+              () => 2,
+              () => {},
+            )
+            onCleanup(stop)
+          },
+          { immediate: true },
+        )
+        return () => ''
+      },
+    }
+    const root = nodeOps.createElement('div')
+    createApp(Comp).mount(root)
+    expect(instance!.scope.effects.length).toBe(3)
+
+    instance!.scope.stop()
+    expect(instance!.scope.effects[0].active).toBe(false)
```

Comment:
I'm wondering why the test specifically checks `effects[0]` and not also `effects[1]`?

I also worry that this is maybe digging a bit too far into the internals, e.g. making assumptions about exactly what effects will be stored in the scope and in what order. Could it maybe be tested via the public API instead?

Reply:
I updated the test case to match the use case of #5783